force-app/main/service/DonationHistoryService.cls

```diff
@@ -88,100 +86,48 @@ public with sharing class DonationHistoryService {
-
-
-    private List<OpportunityContactRole> buildHardCreditHistoryFromOCRs(List<OpportunityContactRole> credits) {
-        List<OpportunityContactRole> hardCreditOCRs = new List<OpportunityContactRole>();
-
-        for(OpportunityContactRole ocr: credits){
-            if (ocr.IsPrimary) {
-                hardCreditOCRs.add(ocr);
-            }
-        }
-        return hardCreditOCRs;
-    }
-
-
-    private List<OpportunityContactRole> buildSoftCreditHistoryFromOCRs(List<OpportunityContactRole> credits) {
-        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group 
-        Filter_Rule__mdt filterRule = CMT_FilterRule_SEL.getFilterRuleByDeveloperName(NPSP_CREDIT_ROLES);
-        List<String> softCreditRoles = filterRule.Constant__c.split(';');
-        List<OpportunityContactRole> softCreditOCRs = new List<OpportunityContactRole>();
-        for(OpportunityContactRole ocr : credits) {
-            if (softCreditRoles.contains(ocr.Role) && !ocr.IsPrimary) {
-                softCreditOCRs.add(ocr);
-            }
-        }
-        return softCreditOCRs;
-    }
-
-
-    private List<OpportunityContactRole> buildPartialSoftCredits(List<OpportunityContactRole> credits) {
-        List<Partial_Soft_Credit__c> partialSoftCreditRecords =
-                selector.getPartialSoftCredits(this.getSoftCreditIds());
-        Set<Id> partialSoftCreditOCRIds = new Set<Id>();
-        for(Partial_Soft_Credit__c partialSoftCreditRecord : partialSoftCreditRecords) {
-            partialSoftCreditOCRIds.add(partialSoftCreditRecord.Contact_Role_ID__c);
+    /**
+     * @descritpion returns the years where the contact made at least 1 donation.
+     * @param contactId the contact id.
+     * @return List<String>
+     */
+    public List<String> getYearsWithDonationsForContact() {
+        
+        if (!hasAccessToDonationHistory()) {
+            throw new UTIL_Permissions.InsufficientPermissionException(Label.commonInsufficientPermissions);
         }
-        List<OpportunityContactRole> partialSoftCreditOCRs = new List<OpportunityContactRole>();
-        for(OpportunityContactRole ocr : credits) {
-            if (partialSoftCreditOCRIds.contains(ocr.Id)) {
-                partialSoftCreditOCRs.add(ocr);
-            }
+        return convertAggregateResultListToStringList(selector.getYearsWithDonationForContact(contactId));
+     }
+
+    /**
+     * @description converts a List<AggregateResult> in a List<String> .
+     * @param result The list with with the aggregate results from the database.
+     * @return List<String> the list of the values in the expr0 represented as String.
+     */ 
+    private List<String> convertAggregateResultListToStringList(List<AggregateResult> results) {
+        List<String> aggregateResultsAsStrings = new List<String>();
+        for(AggregateResult result : results) {
+            aggregateResultsAsStrings.add(String.valueOf(result.get('expr0')));
```

Comment:
Instead of `expr0`, lets give a name to the aggregate expression in the SOQL query. Would 'calendarYear' or 'closeYear' work?